src/rand.rs

```diff
@@ -21,9 +22,100 @@ impl Distribution<Decimal> for Standard {
     }
 }
 
-#[test]
-fn has_random_decimal_instances() {
-    let mut rng = rand::rngs::OsRng;
-    let random: [Decimal; 32] = rng.gen();
-    assert!(random.windows(2).any(|slice| { slice[0] != slice[1] }));
+impl SampleUniform for Decimal {
+    type Sampler = DecimalSampler;
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct DecimalSampler {
+    mantissa_sampler: UniformInt<i128>,
+    scale: u32,
+}
+
+impl UniformSampler for DecimalSampler {
+    type X = Decimal;
+
+    #[inline]
+    fn new<B1, B2>(low: B1, high: B2) -> Self
+    where
+        B1: SampleBorrow<Self::X> + Sized,
+        B2: SampleBorrow<Self::X> + Sized,
+    {
+        let high = *high.borrow();
+        let high = Decimal::from_i128_with_scale(high.mantissa() - 1, high.scale());
+        UniformSampler::new_inclusive(low, high)
+    }
+
+    #[inline]
+    fn new_inclusive<B1, B2>(low: B1, high: B2) -> Self
+    where
+        B1: SampleBorrow<Self::X> + Sized,
+        B2: SampleBorrow<Self::X> + Sized,
+    {
+        let mut low = *low.borrow();
+        let mut high = *high.borrow();
+
+        // Set scales to match one another, because we are relying on mantissas'
+        // being comparable in order outsource the actual sampling implementation.
+        low.rescale(low.scale().max(high.scale()));
```

Comment:
This works perhaps for the common cases, however just to note that it could cause some weird behavior if using one high scale and one large number but low scale. 

e.g. 
`1.000_000_000_000_000_000_01` and ` 100_000_000_000_000_000_001`. In this case, the high would try to rescale however will be unable to. At the moment, the function will not fail and instead return the number scaled to the highest possible precision that it can.

This would of course mean that the mantissa's will be different scales and potentially cause odd behavior (i.e. it'd try to generate random numbers only within the high scale). Consequently, this would not give the range one might expect.

This is a tricky one to circumvent with the current method. One feasible approach is to leverage `trunc` and/or `frac` to generate the two values and then squeeze them in. This also has a few complexities attached with it and won't give a clean `1.00` / `1.01` solution as per below.

Another approach is to instead scale it down to the lowest scale and go that route, however you would also need to be careful since scaling down will round.

I don't have a great answer, however it is something to think about. If this instead returned a `Result` you could potentially handle the "mismatch" scale scenario as an error, however I'm not sure if this is simply a trait implementation or not.

Reply:
> In this case, the high would try to rescale however will be unable to. At the moment, the function will not fail and instead return the number scaled to the highest possible precision that it can.

Thanks, this is a great point that I had not considered.

I assert the goals should be as follows:

- In the common case, the user should get expected behavior, which I assert is "the higher scale".
  - Evidence for my intuition on keeping higher scale, if one wanted a random sample of `dec!(1.0)..=dec!(3.14159)`, it would be expected that values between `3.0` and `3.14159` could actually be yielded.
  - Worse, on the other end: `dec!(1.25)..=dec(3)` could yield the value `1`, which is obviously wrong.
  -  Also, if the user wants the lower scale, rounding your more precise value seems to be more intuitive than rescaling your less precise value.
- Similarly, precision should not be lost wherever possible.

Your point is still valid though; we'll get wild, undefined behavior in the edge case you point out. Here's how I propose we solve it.

1. Rescale both values to the higher value (as we do now).
2. Check and see if the scales are equal; if they are not, rescale to the _lower_ of the two scale values. In other words, we lose precision in this case, but only this case.

This strikes me as a minor "breach of contract" from the goals described above, but it seems to be the best middle ground. It's basically graceful degradation (I also assert that the appetite for this particular use case probably does not exist, although I agree it is right to consider it and try to degrade gracefully. :-))

The code for this is trivial, it just requires adding (immediately after the logic under discussion):

```rs
if low.scale() != high.scale() {
    low.rescale(low.scale().min(high.scale());
    high.rescale(low.scale().min(high.scale());
}
```

What do you think?

> Another approach is to instead scale it down to the lowest scale and go that route, however you would also need to be careful since scaling down will round.

I'd like to avoid doing this as the common case. I could see it being acceptable in the edge case, though.

Reply:
The alternative is to check that the scales are equal and just straight up panic if they aren't.

Reply:
Fixed (using the graceful degradation approach) in 5bbe252.